src/twisted/conch/ssh/keys.py

```diff
@@ -1114,6 +1144,59 @@ def size(self):
             return 256
         return self._keyObject.key_size
 
+    @property
+    def hashAlgorithm(self):
+        """
+        The hash algorithm to use for signing and verifying.
+        """
+        if self._hashAlgorithm is None:
+            keyType = self.type()
+            if keyType in ("RSA", "DSA"):
+                return hashes.SHA1()
+            elif keyType == "EC":
+                # Hash size depends on key size
+                keySize = self.size()
+                if keySize <= 256:
+                    return hashes.SHA256()
+                elif keySize <= 384:
+                    return hashes.SHA384()
+                else:
+                    return hashes.SHA512()
+            elif keyType == "Ed25519":
+                return hashes.SHA512()
+        return self._hashAlgorithm
```

Comment:
Very minor suggestion. I hope it helps with searching for if /elif pairs :)

```suggestion
        if self._hashAlgorithm:
            # We have an explicit algorithm.
            return self._hashAlgorithm

        # When no explicit algorithm is defined we use a default
        # one based on the key type.
        keyType = self.type()
        if keyType in ("RSA", "DSA"):
            return hashes.SHA1()

        if keyType == "EC":
            # Hash size depends on key size
            keySize = self.size()
            if keySize <= 256:
                return hashes.SHA256()
            if keySize <= 384:
                return hashes.SHA384()
            # For any other size go SHA512 as the default.
            return hashes.SHA512()

       if keyType == "Ed25519":
            return hashes.SHA512()
```

Reply:
Thanks - but I think I really prefer the `if`/`elif` chain in this case.  I find that clearer than repeatedly falling through like this.  Sometimes this sort of thing can be worth it to reduce indentation level, but I don't think it makes enough of a difference in this case to be worth (for me) the extra confusion in reading it.